pkg/controller/docdb/dbinstance/dbinstance.go

```diff
@@ -0,0 +1,343 @@
+	if cr.Spec.ForProvider.KMSKeyID != nil && !awsclient.IsEqualString(cr.Spec.ForProvider.KMSKeyID, instance.KmsKeyId) {
+		return false, errors.New(errKMSKeyIDModified)
+	}
+
+	if (cr.Spec.ForProvider.AutoMinorVersionUpgrade != nil && !awsclient.IsEqualBool(cr.Spec.ForProvider.AutoMinorVersionUpgrade, instance.AutoMinorVersionUpgrade)) ||
+		(cr.Spec.ForProvider.CACertificateIdentifier != nil && !awsclient.IsEqualString(cr.Spec.ForProvider.CACertificateIdentifier, instance.CACertificateIdentifier)) ||
+		!awsclient.IsEqualString(cr.Spec.ForProvider.DBInstanceClass, instance.DBInstanceClass) ||
+		!awsclient.IsEqualString(cr.Spec.ForProvider.DBInstanceIdentifier, instance.DBInstanceIdentifier) ||
+		(cr.Spec.ForProvider.PreferredMaintenanceWindow != nil && !awsclient.IsEqualString(cr.Spec.ForProvider.PreferredMaintenanceWindow, instance.PreferredMaintenanceWindow)) ||
+		(cr.Spec.ForProvider.PromotionTier != nil && !awsclient.IsEqualInt64(cr.Spec.ForProvider.PromotionTier, instance.PromotionTier)) {
+		return false, nil
+	}
+
+	areTagsEqual, err := areTagsEqual(e, cr, instance)
+	if err != nil {
+		return false, err
+	}
+
+	if !areTagsEqual {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (e *hooks) preUpdate(ctx context.Context, cr *svcapitypes.DBInstance, obj *svcsdk.ModifyDBInstanceInput) error {
+	if cr.Spec.ForProvider.CACertificateIdentifier != nil {
+		obj.CACertificateIdentifier = cr.Spec.ForProvider.CACertificateIdentifier
+	}
+
+	// obj.DBInstanceIdentifier is set to cr.Spec.ForProvider.DBInstanceIdentifier by default.
+	// However, if the value is changed, external-name should be used to resolve the request.
+	externalName := meta.GetExternalName(cr)
+	if externalName != awsclient.StringValue(cr.Spec.ForProvider.DBInstanceIdentifier) {
+		obj.DBInstanceIdentifier = awsclient.String(externalName)
+		obj.NewDBInstanceIdentifier = cr.Spec.ForProvider.DBInstanceIdentifier
+	}
+
+	// TODO: Should we apply the changes immediately?
+	obj.ApplyImmediately = awsclient.Bool(true)
+
+	return nil
+}
+
+func (e *hooks) postUpdate(ctx context.Context, cr *svcapitypes.DBInstance, resp *svcsdk.ModifyDBInstanceOutput, upd managed.ExternalUpdate, err error) (managed.ExternalUpdate, error) {
+	if err != nil {
+		return managed.ExternalUpdate{}, err
+	}
+
+	respDBIdentfier := awsclient.StringValue(resp.DBInstance.DBInstanceIdentifier)
+	if meta.GetExternalName(cr) != respDBIdentfier {
+		meta.SetExternalName(cr, respDBIdentfier)
+	}
+
+	// TODO: Right now ListTagsForResource needs to be called twice, for isUpToDate and postUpdate
+	// Is there a way to optimize this?
+	errUpdate := updateTags(e, cr, resp.DBInstance)
+	if errUpdate != nil {
+		return managed.ExternalUpdate{}, errors.Wrap(errUpdate, errUpdateTagsFailed)
+	}
+
+	return upd, nil
+}
+
+func (e *hooks) preCreate(ctx context.Context, cr *svcapitypes.DBInstance, obj *svcsdk.CreateDBInstanceInput) error {
+	obj.DBInstanceIdentifier = cr.Spec.ForProvider.DBInstanceIdentifier
```

Comment:
We should always use external-name and ignore `DBInstanceIdentifier` as described [here](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md#external-name).

Reply:
Fixed